Update tuyaha to 0.0.8 and adapt code

ollo69 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR fix Tuya error and implement new features released with last version of tuyaha library.
Fix and features was already tested in [tuya_custom](https://github.com/ollo69/ha_tuya_custom) custom component created some months ago to work-around tuya issue caused by tuya API.
This PR is related to PR #41365
New features will require update to Integration documentation

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #36744, #36713, #37168, #37880, #38438, #39253, #38903, #38156
- This PR is related to issue: PR #41365
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15388

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ollo69 committed:
Merge pull request #1 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #2 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #3 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #4 from home-assistant/dev

Align to master

ollo69 committed:
Merge pull request #5 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #6 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #7 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #8 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #9 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #10 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #11 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #12 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #13 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #14 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #15 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #16 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #17 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #18 from home-assistant/dev

Align from master

ollo69 committed:
Merge pull request #19 from home-assistant/dev

Align from master

ollo69 committed:
Tuya fix and features from tuya_custom

ollo69 committed:
Flake8 fix

ollo69 committed:
Pylint fix

KevinCathcart commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -50,34 +73,90 @@
                     vol.Optional(CONF_PLATFORM, default="tuya"): cv.string,
                 }
-            )
+            ).extend(TUYA_DEVICE_CONF_SCHEMA)
```
Unfortunately, additions/changes to the yaml schema are no longer allowed in integrations for devices or services. See [ADR0010](https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision).

It is not clear to me how one could offer per-device configuration options like these via config flow, given that you also have settings shared between devices (without the latter one could just do one config entry per device). 

Indeed, supporting device level configuration is something that "should be added", per [@balloob in an architecture discussion](https://github.com/home-assistant/architecture/issues/381#issuecomment-625530066).

Given that these options are currently needed to make certain devices work properly, this does need to be exposed. While handling these via entity customization would be possible, that is not permitted to be used for potentially required things like these. Perhaps an exception ADR0010 should be made here, pending such a per-device configuration system being implemented?

I do wonder how the official Tuya app handles things, given that the rest API fails to expose the needed information in all cases. Does the app use some undocumented APIs? Does it have some database of devices built in? Is it simply broken for some devices? (Seems unlikely.) If we could somehow avoid the need to enable users to manually configure these to have the devices work, that would be ideal.

ollo69 replied:
The API that HA via TuyaHA library is using, is a special API that was developed by Tuya some years ago. Recently Tuya decide to limit the access frequency to the call, and this is a creating a lot of issue and error due to fact that the API method is called with very high frequency, especially for user that have a great number of registered device.
Tuya provide an SDK that allow vendor that use tuya Cloud to create dedicated app, but this is not applicable to HA. Tuya and SmartLive app use the same logic, anyway I I have no idea on how this could be applied to HA

Dedicate HA Tuya API also have some limitation that cause some device not working properly because the Cloud do not provide all necessary information, and the configuration.yaml information is used to configured this missing info using some additional method implemented in TuyaHA library with last PR.
In any case I understand that usage of configuration.yaml for new integration should be avoided, but, as you also say, at this moment is not possible to provide configuration for single device using UI.

As alternative approach, I can also remove all device configuration from this PR and implement only the fix that will avoid connection error, the result is that already know limitation for the moment will not be fixed but al least status refresh and control for working device will be fixed.
Please let me know how I should proceed.

KevinCathcart replied:
Yeah, I get all that. I had not noticed this was a special API though. Looking a bit deeper into things they literally have at least 3 different http APIs, which is rather crazy. 

The issues caused by the HA API now having strict rate limits, which can break the integration for people with multiple devices is certainly a high priority thing to merge.

In addition the special HA API fails to return information that is needed, so per device configuration is required to ensure devices can work. Hence the need for per device configuration. This is also an important thing to address.

Personally I do feel like this is something worth making a one-time exception to the ADR for, but that is not my decision to make. Hopefully somebody from the core dev team will comment on this.

bdraco replied:
Rather then implementing additional schema, can this data be collected in a config flow similar to how https://github.com/rospogrigio/localtuya-homeassistant is doing it?

ollo69 replied:
I think that local_tuya use different entry_id for every device, while Tuya integration use a single entry_id for multiple device and new registered device in Tuya app are auto discovered. Furthermore implementing multiple entry_id require a refresh call for every device and this would invalidate the optimization work done, apart from this will completely change the integration behavior.
I could also try to implement an initial wizard for all discovered device during integration configuration, but than will be impossible to configure new discovered device without removing and re-adding integration.

KevinCathcart replied:
Technically, one can handle this during editing too. But this is not going to be terribly ergonomic.

To lay out such an approach to be as general as possible (possibly more general than would be needed here):
1. First form(s) is/are normal shared level options.
2. Next form is a menu form that asks what additional action to take: "Add Device Configuration", "Remove Device Configuration", "Edit Device Configuration", "Finished Configuring".
     1. If you pick add, it takes you to a form with device level settings. And when finished, it returns you to the menu form.
     2. If you pick "edit" it shows you a list of devices that currently have configuration. You pick one, it brings you to a form to edit it. Once done, you return to the menu.
     3. If you pick "Remove", you see a list of devices that currently have configuration. You can pick one or more. When you click submit it takes you back to the menu.
     4. If you pick "Finished Configuring", the process ends.

Obviously in between each step you are applying the relevant changes to the in-progress version of the config entry.

One could probably do slightly better if discovered devices are available. In that case, one would know what all the discovered devices are, and could simply list them, along with options for "edit" or "reset". Edit would create a default configuration for said device it it did not have one already, and then show an edit dialog. Reset would simply delete the device level configuration if one exists.

Where things get really messy is if a device configuration needs some form of list of complex values. One would probably need to repeat a similar system menu system at the device level, nested inside the other menu.

ollo69 replied:
I will try do do something as suggested. I will report progress...

ollo69 replied:
Working on this, I think that adding a combo in ConfigOption flow I can manage this and remove the additional schema. Need some days to complete this change...

ollo69 replied:
Finish implementing, this what I did:

1) when you have a configurable device (light or climate) you can select in from a drop-down list (default is save configuration)

![image](https://user-images.githubusercontent.com/60491700/96335023-8562e100-1075-11eb-9de8-388dc6ab33bc.png)

3) If you select a device, open the page for the corresponding type:

Light:

![image](https://user-images.githubusercontent.com/60491700/96335049-bf33e780-1075-11eb-9896-f5d136b32900.png)

Climate:

![image](https://user-images.githubusercontent.com/60491700/96335060-de327980-1075-11eb-8a76-02dcd5918656.png)

3) When you press "save", it come back to the main page, so that you can choose to configure another device or finally save configuration leaving default value "Save Configuration" in drop-down.

What do you think?

estebanz01 replied:
I don't have a horse in this race, but I'd rename `Save configuration` to something that tells me it's going to apply across devices and not per device.

ollo69 replied:
Suggestion?

estebanz01 replied:
what about `All devices` ? it aligns with the message `Select device ...`

ollo69 replied:
But `Save Configuration` means that you are going to "Save the configuration" when you press "Apply". `All device` will not clarify this. May be I do not understand what you have in mind?

estebanz01 replied:
So you have an option that says: `Select the device to configure` and in the dropdown, you have two options: `Save configuration` and `Lampux-RGBCWA60 (light)`. I was thinking that is better to rename the dropdown option to `All devices`, so when you click done, you know it applies to _all_. 

Not sure if that helps my suggestion more clear 😅

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +289,88 @@ async def async_unload_entry(hass, entry):
+        dev_type = self._tuya.device_type()
+        if any(c in dev_type for c in ["scene", "switch"]):
+            return
+        if TuyaDevice._device_count > 0:
```
Why is this check needed?

ollo69 replied:
Just a check to be sure that device count don't go below 0, but probably not really useful.

bdraco replied:
It seems like it would be a bug if it decremented the device count too far.  Ideally we would have a test for this instead of a guard in the code.

ollo69 replied:
Can you better clarify what you mean?

ollo69 replied:
Removed check, unuseful

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -187,9 +232,63 @@ def supported_features(self):
+            if len(attr_list) > 2:
+                entity_attr = attr_list[2]
+
+            state_obj = self.hass.states.get(entity_name)
```
This looks like its re-implementing part of `generic_theromostat` https://www.home-assistant.io/integrations/generic_thermostat/

Is there no way to get the temperature from the tuya api for the climate device?

ollo69 replied:
Temperature come from device by default, but some device do not provide this info. So I add the possibility to configure an external sensor to provide the missing information. This is one of the additional configuration option in the new schema.

bdraco replied:
Is the external sensor provided by the integration or another integration?

ollo69 replied:
Any sensor in HA.

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -105,13 +188,26 @@ async def async_setup_entry(hass, entry):
+        }
+    )
+
+    _update_discovery_interval(
```
Since this is rate limited on the tuya side, I'm not sure it makes sense to give the user control of the interval since it will most likely be ineffective below the defaults.

ollo69 replied:
Not sure that Tuya will not change this limit again in the future, hoping in lower values. In any case if the limit is set to low, a warning message is write in the log, so this option allow a tuning of the integration without requiring to modify the code.

bdraco replied:
Generally we don't add options to change these unless there is an exceptional case.  Unless there is a history of the provider changing this frequently, we should set this to the fastest allowed polling interval allowed that doesn't flood the state machine with updates.

ollo69 replied:
Current allowed interval is very high (600 sec.). They change it within a month from 300 to 600. Lower limit that can be set now by integration option is 10 seconds, I can set to higher value that make sense for you. I think that leaving the option to reduce current interval inside a limited range should be better...

ollo69 replied:
I increased the lower limit for this options to 30 seconds

estebanz01 commented:
not sure if this is the right place to comment @ollo69 but feel free to ping me for tuya QA and some code help if needed. I have a bit more time to work on my HA instance now 😉

ollo69 committed:
Added device options inside config_flow

ollo69 committed:
Fixed iSort

ollo69 committed:
Fixed pylint

ollo69 committed:
Strings changed

ollo69 committed:
Increased discovery and query interval options lower limit to 30 seconds

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -89,21 +108,57 @@ def __init__(self, tuya, platform):
         """Create operation list when add to hass."""
         await super().async_added_to_hass()
+        self._load_config()
+        async_dispatcher_connect(self.hass, SIGNAL_CONFIG_ENTITY, self._load_config)
```
Wrap this in self.async_on_remove

ollo69 replied:
What do you mean? Unsubscribe dispatcher on remove?

bdraco commented on homeassistant/components/tuya/light.py:
```diff
@@ -64,6 +86,45 @@ def __init__(self, tuya, platform):
+        """Set config parameter when add to hass."""
+        await super().async_added_to_hass()
+        self._load_config()
+        async_dispatcher_connect(self.hass, SIGNAL_CONFIG_ENTITY, self._load_config)
```
This one as well

bdraco reviewed (changes requested):
I'll test out the config flow as soon as I have some free cycles

bdraco commented:
@estebanz01 have you had a chance to test this yet?

estebanz01 commented:
@bdraco no yet. I'll set the environment today.

I have 6 lights, one cover and a bunch sockets. I will report later in the day.

estebanz01 commented:
Config flow is amazing! 🦜

estebanz01 commented:
OK, some testing with my lights and poll interval. I found two things:

* What's the minimum for status retrieval? The range slider is not consistent with the right range number, so it is a bit confusing because you can keep decreasing the number while the range is at the minimum 😄 .
![Captura de pantalla 2020-10-18 a la(s) 10 44 31 a  m](https://user-images.githubusercontent.com/2417465/96372968-311c3780-112f-11eb-865b-5940563583aa.png)

* I changed the status interval check to 60 secs and I'm getting timeouts. This would indicate that tuya is throttle me, but that's OK. The problem is that the status in lovelace is not being reflected, but I see in the history log of the device that the status change has been acknowledge.

```python
Traceback (most recent call last):
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/homeassistant/helpers/entity.py", line 278, in async_update_ha_state
    await self.async_device_update()
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/homeassistant/helpers/entity.py", line 482, in async_device_update
    await task
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/homeassistant/components/tuya/__init__.py", line 362, in update
    self._tuya.update(use_discovery=(TuyaDevice._device_count > 1))
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/tuyaha/devices/base.py", line 134, in update
    return self._update(use_discovery)
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/tuyaha/devices/base.py", line 84, in _update
    devices = self.api.discovery()
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/tuyaha/tuyaapi.py", line 188, in discovery
    response = self._request("Discovery", "discovery")
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/tuyaha/tuyaapi.py", line 245, in _request
    response = self._requestSession.post(
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/requests/sessions.py", line 578, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/Users/esteban/homeassistant/lib/python3.8/site-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='px1.tuyaus.com', port=443): Read timed out. (read timeout=None)
```

![Captura de pantalla 2020-10-18 a la(s) 10 49 34 a  m](https://user-images.githubusercontent.com/2417465/96373051-a7b93500-112f-11eb-9364-36b51cbc5a0c.png)

but so far, this looks _A M A Z I N G_. Love it!

ollo69 commented:
> OK, some testing with my lights and poll interval. I found two things:
> 
> * What's the minimum for status retrieval? The range slider is not consistent with the right range number, so it is a bit confusing because you can keep decreasing the number while the range is at the minimum 😄 .

But this is not related to component, may be should be opened a ticket to the lovelace team.

>I changed the status interval check to 60 secs and I'm getting timeouts. This would indicate that tuya is throttle me, but that's OK. The problem is that the status in lovelace is not being reflected...

This 2 parameters are here for tuning, the idea is to be able to decrease the polling interval in case that Tuya will provide better refresh rate (also not sure that allowed refresh rate is equal for any region), but if you set a too low value you will start having error in the log.
But my concern, also reading your comment, is that this will not be correctly interpreted and users will simply set this to minimum value expecting corresponding result. 
I have also to keep in mind that if I remove this 2 options and there are not configurable devices (light or climate), I will have an option button in integration page that will do nothing (not sure that is possible hide option button programmatically).

> ... but I see in the history log of the device that the status change has been acknowledge.

Your status changed after 10 minutes, as expected. If you better read your log, you will see that the error is every 60 seconds, but after 600 seconds probably it disappear and than come back again every 60 sec.

> ```python
>....
> requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='px1.tuyaus.com', port=443): Read timed out. (read timeout=None)
> ```

This is another strange thing! In my region when Tuya is polled below a specific interval, it return a specific API message ("Frequently Invoke") that is intercepted and raised as specific Exception by TuyaHA. In your case seems different, because you are receiving a ReadTimeout exception. For sure I have to catch also this case.

estebanz01 commented:
@ollo69 I think the mismatch between the toggles in lovelace is a minor thing and I'm being picky about it. Regarding range within the options view, I believe they are useful. Maybe it needs a bit more explanation that can be addressed in the docs, no ? I think it's better to have it there regardless of the device than just having it for devices that allow configuration.

ollo69 commented:
There are some example or reference on how I should create documentation for options?

ollo69 committed:
Minor change

- added options dispather unsubscription on async remove
- changed descriptions strings
- changed log level for Tuya connecton error

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -89,21 +108,59 @@ def __init__(self, tuya, platform):
         """Create operation list when add to hass."""
         await super().async_added_to_hass()
+        self._load_config()
+        self._unsub_dispatcher = async_dispatcher_connect(
+            self.hass, SIGNAL_CONFIG_ENTITY, self._load_config
+        )
+
```
```suggestion
        self.async_on_remove(async_dispatcher_connect(
            self.hass, SIGNAL_CONFIG_ENTITY, self._load_config
        ))

```

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +242,94 @@ async def async_unload_entry(hass, entry):
+
+    async def async_will_remove_from_hass(self):
+        """Unregister config update dispatcher."""
+        if self._unsub_dispatcher is not None:
+            self._unsub_dispatcher()  # pylint: disable=not-callable
+            self._unsub_dispatcher = None
```
```suggestion
```

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +242,94 @@ async def async_unload_entry(hass, entry):
         """Init Tuya devices."""
         self._tuya = tuya
         self._tuya_platform = platform
+        self._unsub_dispatcher = None
```
```suggestion
```

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +242,94 @@ async def async_unload_entry(hass, entry):
-        dev_id = self._tuya.object_id()
-        self.hass.data[DOMAIN]["entities"][dev_id] = self.entity_id
+        self.hass.data[DOMAIN]["entities"][self.object_id] = self.entity_id
         async_dispatcher_connect(self.hass, SIGNAL_DELETE_ENTITY, self._delete_callback)
         async_dispatcher_connect(self.hass, SIGNAL_UPDATE_ENTITY, self._update_callback)
```
```suggestion
        self.async_on_remove(async_dispatcher_connect(self.hass, SIGNAL_DELETE_ENTITY, self._delete_callback))
        self.async_on_remove(async_dispatcher_connect(self.hass, SIGNAL_UPDATE_ENTITY, self._update_callback))
```

ollo69 committed:
Added call to async_on_remove for dispatcher unsub

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -107,12 +144,24 @@ async def async_setup_entry(hass, entry):
         ENTRY_IS_SETUP: set(),
         "entities": {},
         "pending": {},
+        entry.entry_id: {
```
Since we only allow one entry, its ok to store "listener" without the `entry.entry_id`

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +242,98 @@ async def async_unload_entry(hass, entry):
 
+    def _inc_device_count(self):
+        """Increment static variable device count."""
+        dev_type = self._tuya.device_type()
+        if dev_type in ["scene", "switch"]:
+            return
```
Since this code is the same as the code in `_dec_device_count`, how about breaking this out into something like `self._device_uses_discovery`?? I'm not sure that name makes sense as I'm not 100% clear on the intent here.

ollo69 replied:
Called `self._device_can_use_query`

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +242,98 @@ async def async_unload_entry(hass, entry):
+        """Get updated device options."""
+        dev_conf = {}
+        devices_config = self.hass.data[DOMAIN].get(TUYA_DEVICES_CONF)
+        if devices_config:
```
```suggestion
        if not devices_config:
            return {}
```

Then outdent the block below

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -89,21 +108,59 @@ def __init__(self, tuya, platform):
             return
 
         for mode in modes:
             if mode in TUYA_STATE_TO_HA:
```
```suggestion
            if mode not in TUYA_STATE_TO_HA:
                continue
```

Then outdent the block below

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -187,9 +252,61 @@ def supported_features(self):
+                self._temp_entity_error = True
+
+        try:
+            attr_list = self._temp_entity.split(".", 2)
+            if len(attr_list) < 2:
+                _log_error("entity name is invalid")
+                return None
```
Validate this at configure time instead so we don't let the invalid data into the config.

ollo69 replied:
Should I check also that sensor exist or only the syntax?

bdraco replied:
Since its in the config flow we expect HA to already be started so it seems fine check for existence as well.  Ideally you could borrow the code from `homekit`s config flow and allow them to choose the entity in the config flow so they don't mistype it.

We should keep the existence check here as well in case it disappears though

ollo69 replied:
I did it, anyway this is limiting a little bit the options compared with the original implementation where was possible to also select an entity attribute instead of the attribute state. In any case I think that this will avoid a lot of typing errors and so it is acceptable.

bdraco replied:
Home Assistant is moving away from having sensors as additional attributes.  If the integration hasn't been updated yet, it can be wrapped with a template sensor.

bdraco commented on homeassistant/components/tuya/cover.py:
```diff
@@ -60,6 +62,8 @@ def __init__(self, tuya, platform):
         """Init tuya cover device."""
         super().__init__(tuya, platform)
         self.entity_id = ENTITY_ID_FORMAT.format(tuya.object_id())
+        self._was_closing = False
+        self._was_opening = False
```
Can we move this into the pypi package?

ollo69 replied:
Sorry, I don't understand what you mean.

bdraco replied:
Ideally the library returns the state of the cover so we don't have `self._was_closing = False`  and `self._was_opening = False` in Home Assistant.

Maybe `self._tuya.cover_state()`

Maybe an enum with the following states.

TuyaCoverState.CLOSING
TuyaCoverState.CLOSED
TuyaCoverState.OPENING
TuyaCoverState.OPENED

ollo69 replied:
Tuya API do not provide closing and opening, so this state here is just deducted based on the command send.
I could also open a PR in TuyaHA, but normally code owner take a long time to manage this...

bdraco replied:
Unfortunately we aren't allowed to fake state in Home Assistant.  Please take this out and move it into the pypi package.  I will prioritize a follow up review when the package owner gets the merge done.

ollo69 replied:
Wait, I think that I do not properly explain, because I too had forgotten the correct logic.

Tuya provide 3 states:
- 1=Opening
- 2=Closing
- 3=Stopped

What is missing is closed and open, so I was using the 2 variable to implement this logic: 
- if status != 2 and previous was 2 (_was_closing) -> cover is closed (is_closed return true)
- if status != 1 and previous was 1 (_was_opening) -> cover is open (is_closed return false)
- other case return none (means not fully open nor closed)

When cover is manually stopped, the 2 variable related to previous state are reset.

How do you think this 3 state should differently be managed?

bdraco replied:
If the state isn't polled while the cover is opening or closing will the state be incorrect?

ollo69 replied:
It will be None because we will only know that is not moving.

bdraco replied:
I ordered one of these for testing.

https://www.amazon.com/Controller-Anywhere-Compatible-Assistant-Required/dp/B07PRRY5NN

bdraco replied:
Edit: I found one I can get tomorrow https://www.amazon.com/gp/product/B07VWNCR14/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1

bdraco replied:
Garage door opener came in, but the integration doesn't find it.

ollo69 replied:
Unfortunately this may happen, not all device are supported by this API.

ollo69 replied:
If can help, you can refer to [this](https://github.com/ollo69/ha_tuya_custom/issues/7) discussion.
But if this check is an issue, I can finally remove it and always return `None`  from `is_closed` method, that means that returned status will be `Opening`, `Closing `or `Unknown`

bdraco replied:
At first I was hitting the rate limit.  It turns out that calling `tuya.get_access_token()` appears to reset the rate limit at least for the discover call.

Once I got past that, the garage door device was missing from the endpoint.

bdraco replied:
If the `cover` device is something like this https://www.amazon.com/Essenc-Curtain-Blinds-Shutter-Control/dp/B08LPFCTGV/ref=sr_1_15?dchild=1&keywords=tuya+curtain&qid=1603682860&sr=8-15 the state should be `None` (which will be UNKNOWN) and we only trigger open/close and don't know the state.

It looks like we allowed an optimistic mode https://github.com/home-assistant/core/pull/31407 so something like that should be ok as well.

estebanz01 replied:
the cover is something like one of this two products: https://www.amazon.com/Rollerhouse-Tubular-Rechargeable-Electric-Curtains/dp/B078XDZD1B/ref=sr_1_3?crid=2ME291PH11QNB&dchild=1&keywords=tuya+curtain+motor&qid=1603684005&sprefix=tuya+curt%2Caps%2C247&sr=8-3 or https://www.ajaxonline.co.uk/product/smart-blind-motor/

I have one, so let me know if there's something I need to test.

bdraco replied:
It looks like https://www.amazon.com/Electric-Control-Drapery-compatiable-Broadlink/dp/B07SWRQ7GX/ref=sr_1_2?dchild=1&keywords=tuya+curtain&qid=1603763564&sr=8-2 
or
https://www.amazon.com/Motorized-Roller-36-38MM-Shutter-Control/dp/B0836J7QT9

could be the device as well so its probably ok as-is.

I'd order the second one for testing but I'm about to be traveling so I wouldn't get it in time.  Given the integration is barely working now due to the rate limits, it seems like something we can adjust later if need be.

bdraco commented on homeassistant/components/tuya/light.py:
```diff
@@ -64,6 +86,47 @@ def __init__(self, tuya, platform):
+    async def async_added_to_hass(self):
+        """Set config parameter when add to hass."""
+        await super().async_added_to_hass()
+        self._load_config()
```
```suggestion
        self._process_config()
```

Can we avoid calling this `load` since its in `async` and as it implies we are doing I/O

ollo69 replied:
Renamed also in `climate.py`

bdraco reviewed (changes requested):
I've ordered some more tuya devices so I can do testing.  I'll report back once I've had a chance to do a full run though.  In the mean time there are a few minor issues I've commented on.

ollo69 committed:
Implementation of requested changes

ollo69 committed:
Fix for flake8

ollo69 committed:
Rename static attribute

sdrasmussen commented:
I am not a programmer, and know NOTHING about what it takes to make this work, however, I do have currently 53 Tuya devices in Home Assistant. If sharing my logs would help in any way, I'd be glad to. I am running the regular integration, not the custom.

ollo69 committed:
Changed Option Config flow

- added possibility to configure multiple device of the same type
- added sensors list in climate config

ollo69 commented:
@estebanz01, could you make a new test? I changed a little bit the options flow allowing multiple devices configuration., keeping in mind that some users have a very high number of devices. But I have just 1 :)

ollo69 committed:
Minor description_placeholders change

estebanz01 commented:
@ollo69 I'll do some testing later in the day. 👌

ollo69 committed:
Added comments

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +240,100 @@ async def async_unload_entry(hass, entry):
     if unload_ok:
+        hass.data[DOMAIN]["listener"]()
+        hass.data[DOMAIN]["listener"] = None
         hass.data[DOMAIN][ENTRY_IS_SETUP] = set()
```
```suggestion
```
Since you are going to delete it below anyways

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -191,31 +240,100 @@ async def async_unload_entry(hass, entry):
+        hass.data[DOMAIN]["listener"] = None
         hass.data[DOMAIN][ENTRY_IS_SETUP] = set()
         hass.data[DOMAIN][TUYA_TRACKER]()
         hass.data[DOMAIN][TUYA_TRACKER] = None
         hass.data[DOMAIN][TUYA_DATA] = None
+        hass.data[DOMAIN].pop(TUYA_DEVICES_CONF)
```
```suggestion
```

Since you are going to delete it below in the pop

bdraco commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -106,3 +144,237 @@ async def async_step_user(self, user_input=None):
+        """Return configuration form for devices."""
+        count = 0
+        conf_devs_id = []
+        for dev_id in devs_id:
```
```suggestion
        for count, dev_id in enumerate(devs_id):
```

bdraco commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -106,3 +144,237 @@ async def async_step_user(self, user_input=None):
+
+    async def _async_device_form(self, devs_id):
+        """Return configuration form for devices."""
+        count = 0
```
```suggestion
```

bdraco commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -106,3 +144,237 @@ async def async_step_user(self, user_input=None):
+                self._form_error = ERROR_DEV_MULTI_TYPE
+                return await self.async_step_init()
+            conf_devs_id.append(device_info[1])
+            count += 1
```
```suggestion
```

bdraco commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -106,3 +144,237 @@ async def async_step_user(self, user_input=None):
+                    default=self.config_entry.options.get(
+                        CONF_DISCOVERY_INTERVAL, DEFAULT_DISCOVERY_INTERVAL
+                    ),
+                ): vol.All(vol.Coerce(int), vol.Clamp(min=30, max=3600)),
```
```suggestion
                ): vol.All(vol.Coerce(int), vol.Clamp(min=30, max=150)),
```

3600 seems high. Is there an actual use case for this?

ollo69 replied:
600 is too low. Current default value (without error) is 605. I agree that 3600 is very high, changed to 900

bdraco commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -106,3 +144,237 @@ async def async_step_user(self, user_input=None):
+                    default=self.config_entry.options.get(
+                        CONF_QUERY_INTERVAL, DEFAULT_QUERY_INTERVAL
+                    ),
+                ): vol.All(vol.Coerce(int), vol.Clamp(min=30, max=3600)),
```
```suggestion
                ): vol.All(vol.Coerce(int), vol.Clamp(min=30, max=600)),
```

ollo69 replied:
For query method current working value is 120, so I changed max to 240

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -187,9 +256,61 @@ def supported_features(self):
+                self._temp_entity_error = True
+
+        try:
+            attr_list = self._temp_entity.split(".", 2)
```
use `from homeassistant.core import split_entity_id, valid_entity_id`

ollo69 replied:
Reviewed method, removed attribute check and selection

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -68,6 +77,30 @@
+    if not tuya:
+        return
+
+    try:
+        tuya.discovery_interval = float(interval)
+        _LOGGER.info("Tuya discovery device poll interval set to %s seconds", interval)
+    except ValueError as ex:
+        _LOGGER.warning(ex)
```
We shouldn't need to validate that its a float as it should be validated before we accept or store the input

ollo69 replied:
Not validating but just converting, anyway I removed the conversion. Try - except is because attribute setter in library can raise ValueError.

bdraco commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -68,6 +77,30 @@
+    if not tuya:
+        return
+
+    try:
+        tuya.query_interval = float(interval)
+        _LOGGER.info("Tuya query device poll interval set to %s seconds", interval)
+    except ValueError as ex:
+        _LOGGER.warning(ex)
```
We shouldn't need to validate that its a float as it should be validated before we accept or store the input

ollo69 replied:
Not validating but just converting, anyway I removed the conversion. Try - except is because attribute setter in library can raise ValueError.

bdraco reviewed (changes requested):
I ran though some testing with the new devices I picked up, and everything appears to be working as expected.  At this point, I think its just code cleanups remaining

ollo69 committed:
Requested changes

ollo69 commented:
Working on documentation.

ollo69 committed:
Fix pylint

estebanz01 commented:
I got sidetracked last night. I'll start now with testing.

estebanz01 commented:
@ollo69 @bdraco after playing with this PR, everything looks good on 🦜 . My previous concerns were solved and everything works as it should. 🎉 🎉 .

bdraco commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -106,3 +144,237 @@ async def async_step_user(self, user_input=None):
+async def _get_entities_matching_domains(hass, domains):
+    """List entities in the given domains."""
+    included_domains = set(domains)
+    entity_ids = [state.entity_id for state in hass.states.async_all(included_domains)]
```
```suggestion
    entity_ids = hass.states.async_entity_ids(included_domains)
```

ollo69 committed:
Requested changes

ollo69 committed:
Added option to select device that will use query method

ollo69 commented:
Sorry for this last minute change, but yesterday night I had the idea that could be useful to give the possibility to choose the device that will use query method (can be only one but allows better refresh interval). I also hide the query polling interval setting if there are no device that can use query. I'm going to update documentation too.

estebanz01 commented:
testing new changes.

estebanz01 commented:
Working as expected and it's more clear now the options menu without the `Save configuration` option 😉

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -187,9 +256,54 @@ def supported_features(self):
+
+            state_obj = self.hass.states.get(entity_name)
+            if state_obj:
+                temp = state_obj.state
```
in `homeassistant.core.State` `self.state` is a `str`

ollo69 replied:
Fixed

ollo69 committed:
Requested change - Fix status check

bdraco commented on homeassistant/components/tuya/climate.py:
```diff
@@ -187,9 +256,53 @@ def supported_features(self):
+        if not self._temp_entity or self._temp_entity == ENTITY_MATCH_NONE:
+            return None
+
+        def _log_error(error_msg):
+            if not self._temp_entity_error:
+                _LOGGER.warning(
+                    "Error on Tuya external temperature sensor %s: %s",
+                    self._temp_entity,
+                    error_msg,
+                )
+                self._temp_entity_error = True
```
It would be a bit faster to make this `self._log_temp_error` since every time this is called it has to recreate the function.

ollo69 replied:
Done

ollo69 committed:
Fixed flake8

bdraco commented on homeassistant/components/tuya/config_flow.py:
```diff
@@ -106,3 +146,261 @@ async def async_step_user(self, user_input=None):
+        if user_input is not None:
+            dev_ids = user_input.get(CONF_LIST_DEVICES)
+            if dev_ids:
+                return await self._async_device_form(dev_ids)
```
We aren't allowed to show the form for another step.  `_async_device_form` currently displays the form for the `device` step.  

Please move to the `device` step by calling `await self.async_step_device(...)` before showing the form.

ollo69 replied:
Done

ollo69 committed:
Requested changed

bdraco reviewed (approved):
Thanks for doing the update and working though the code review 💯 

Another reviewer may have additional comments.  If so, they should be addressed in a followup PR.

estebanz01 commented:
🎉 🎉 🎉 🎉

MartinHjelmare commented on homeassistant/components/tuya/climate.py:
```diff
@@ -187,9 +255,55 @@ def supported_features(self):
+            )
+            self._temp_entity_error = True
+
+    def _get_ext_temperature(self):
```
We don't allow returning temperature from another entity in the climate entity. That can be done in a template integration eg, but not in a device integration.

Please remove this feature.

MartinHjelmare reviewed (commented):
Please address the comment in a new PR. Thanks!

pergolafabio commented:
is it now merged to 117.3 ?

mfabiani53 commented:
> is it now merged to 117.3 ?

It's not merged also to 0.117.4

ollo69 commented:
This is a big change, probably will be released with next HA version. Please be patient.
